src/gcp_bigquery.rs

```diff
@@ -0,0 +1,449 @@
+///
+/// * `abi_type` - the ABI type, specified as a string
+fn solidity_type_to_db_type(abi_type: &str) -> &str {
+    match abi_type {
+        "address" => "string",
+        "bool" | "bytes" | "string" | "int256" | "uint256" => "string",
+        "uint8" | "uint16" | "uint32" | "uint64" | "uint128" | "int8" | "int16" | "int32"
+        | "int64" | "int128" => "int",
+        _ => panic!("Unsupported type {}", abi_type),
+    }
+}
+
+impl GcpBigQueryClient {
+    ///
+    /// Creates a new GcpBigQueryClient, which includes the connection to the GCP bigquery instance
+    /// As well as project / dataset settings.
+    /// Instance methods encapsulate all utilized behavior (create, delete, insertion/sync)
+    ///
+    /// # Arguments
+    ///
+    /// * indexer_gcp_bigquery_config: bigquery configurations
+    /// * indexer_contract_mappings: contract event mappings
+    ///
+    pub async fn new(
+        indexer_gcp_bigquery_config: &IndexerGcpBigQueryConfig,
+        indexer_contract_mappings: &[IndexerContractMapping],
+    ) -> Result<Self, GcpClientError> {
+        let client = gcp_bigquery_client::Client::from_service_account_key_file(
+            indexer_gcp_bigquery_config.credentials_path.as_str(),
+        )
+        .await;
+
+        let table_map = load_table_configs(indexer_contract_mappings);
+
+        match client {
+            Err(error) => Err(GcpClientError::BigQueryError(error)),
+            Ok(client) => Ok(GcpBigQueryClient {
+                client,
+                drop_tables: indexer_gcp_bigquery_config.drop_tables,
+                project_id: indexer_gcp_bigquery_config.project_id.to_string(),
+                dataset_id: indexer_gcp_bigquery_config.dataset_id.to_string(),
+                table_map,
+            }),
+        }
+    }
+
+    ///
+    /// Deletes tables from GCP bigquery, if they exist
+    /// Tables are only deleted if the configuration has specified drop_table
+    pub async fn delete_tables(&self) {
+        if self.drop_tables {
+            for table_name in self.table_map.keys() {
+                let table_ref = self
+                    .client
+                    .table()
+                    .get(
+                        self.project_id.as_str(),
+                        self.dataset_id.as_str(),
+                        table_name.as_str(),
+                        None,
+                    )
+                    .await;
+                match table_ref {
+                    Ok(table) => {
+                        // Delete table, since it exists
+                        println!("deleting table: {table_name}");
+                        let _ = table.delete(&self.client).await;
+                    }
+                    Err(..) => {}
+                }
+            }
+        }
+    }
+
+    ///
+    /// Iterates through all defined tables, calls create_table on each table
+    pub async fn create_tables(&self) {
+        for (table_name, column_map) in self.table_map.iter() {
```

Comment:
is it important for them to block each other on creation? or could they be created concurrently? not sure if it matters or saves anything speed wise (in a sense you wouldn't notice as its already fast) but just wondering

Reply:
hmm, not sure.  it might be worth experimenting with.  i wonder if there can be concurrent table creations with a single connection?   i imagine it should be okay though.  but to your point above it prob won't add any noticeable latency as its just run once rather than on a per file / batch or row basis.